### PR TITLE
Review PR

### PR DESCRIPTION
**Pull request by MasonEgger**
Review branch for Temporal 101 in Python. Please leave your comments inline

### CODE REVIEW

File: exercises/finale-workflow/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+Since the Activity is written in Java, it's able to use a Java graphics
+library that would otherwise be would be incompatible with a typical 
+Python program. Since the Workflow and Activities are implemented in two 
+different languages, you'll need to start two Worker programs, one 
+written in Java that will execute the Activity code and one written 
+in Python that will execute the Workflow code.
+
+
+
+# Run the Activity Worker (implemented in Java):
```
Comment (cretz) [inline exercises/finale-workflow/README.md:19]: I think you'd get more mileage implementing in Go. It's often easier for people and is a good cross-language target.

File: exercises/finale-workflow/workflow.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+from datetime import timedelta
+from temporalio import workflow
+
+
+@workflow.defn
+class CertificateGeneratorWorkflow:
+    @workflow.run
+    async def run(self, name: str) -> str:
+        out_file_path = await workflow.execute_activity(
+            "CreatePdf", name, start_to_close_timeout=timedelta(seconds=5)
```
Comment (cretz) [inline exercises/finale-workflow/workflow.py:10]: If you wanted to go all out, you could make a dummy activity or protocol to represent the types of the Java side here

**Review by cretz — COMMENTED** [review]
Only a few notes, made most on the other PR

File: exercises/finale-workflow/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+$ python worker.py
+```
+
+# Start the Workflow (Python)
+In another terminal, run the following command,
+but be sure to replace `Mason Egger` with your
+full name.
+
+```
+$ python starter.py "Mason Egger"
```
[the remark this answers is not in the document]
Comment (MasonEgger) [reply exercises/finale-workflow/README.md:42]: :P

File: exercises/finale-workflow/README.md (hunk 1 above)
Comment (MasonEgger) [reply exercises/finale-workflow/README.md:19]: Unfortunately the PDF library necessary isn't available in Go. This was a decision made by Tom when writing the first 101 course.

File: exercises/farewell-workflow/practice/translate.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import requests
+import urllib.parse
+from typing import Tuple
+from temporalio import activity
+
+
+@activity.defn
+async def greet_in_spanish(name: str) -> str:
+    greeting = call_service("get-spanish-greeting", name)
```
[the remark this answers is not in the document]
Comment (MasonEgger) [reply exercises/farewell-workflow/practice/translate.py]: Fixed in [5265990](https://github.com/temporalio/edu-101-python-code/pull/2/commits/5265990a7a0e5d13bf7dd6d8e64bc0b3604c6728)

File: demos/service-workflow/translate.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+from temporalio import activity
+
+
+@activity.defn
+async def greet_in_spanish(name: str) -> str:
+    base = "http://localhost:9999/get-spanish-greeting?name=%s"
+    url = base % urllib.parse.quote(name)
+
+    try:
+        resp = requests.get(url)
```
[the remark this answers is not in the document]
Comment (MasonEgger) [reply demos/service-workflow/translate.py:13]: Fixed in [5265990](https://github.com/temporalio/edu-101-python-code/pull/2/commits/5265990a7a0e5d13bf7dd6d8e64bc0b3604c6728)

File: exercises/finale-workflow/workflow.py (hunk 2 above)
Comment (MasonEgger) [reply exercises/finale-workflow/workflow.py:10]: I think that's a good idea to dive into in a later course

**Comment by cretz** [comment]
@MasonEgger - let me know if/when this is ready for re-review

**Comment by MasonEgger** [comment]
@cretz This is ready for a re-review. However I don't know if much changed that you don't already know about via our Slack conversations. But feel free to do another pass if you like.

File: demos/service-workflow/translate.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cretz) [reply demos/service-workflow/translate.py:13]: Looks like this one didn't get fixed

File: exercises/finale-workflow/README.md (hunk 1 above)
Comment (cretz) [reply exercises/finale-workflow/README.md:19]: A bit rough to ask users to execute opaque compiled code and to ship around a JAR with samples, but meh

**Review by cretz — APPROVED** [review]
LGTM (I just skimmed since I think I reviewed most last time, will wait for content update for bigger review)

File: exercises/farewell-workflow/practice/translate.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+        async with self.session.get(url) as response:
+            translation = await response.text()
+
+            if response.status >= 400:
+                raise ApplicationError(
+                    f"HTTP Error {response.status}: {translation}",
+                )
+
+            return translation
```
Comment (cretz) [inline exercises/farewell-workflow/practice/translate.py:34]:
```suggestion
            response.raise_for_status()
            return await response.text()
```

If you're not gonna reasonably differentiate retryable from non-retryable, you might as well just use the helper and make this code look really simple (and all other code where you do this). You don't really need the body or an `ApplicationError` for failures in your sample.

File: demos/service-workflow/translate.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cretz) [reply demos/service-workflow/translate.py:13]: Bump

[the remark this answers is not in the document]
Comment (MasonEgger) [reply demos/service-workflow/translate.py]: Thanks for the reminder. I forgot about this. Fixed now.

**Review by cretz — APPROVED** [review]

**Comment by MasonEgger** [comment]
This has been finished and published. Closing as it was a main -> review merge so the changes are already live

**Pull request closed without merging**